### PR TITLE
feat: Add PromptTemplate type

### LINKED ISSUE #5771: [hub] [gql] ChatPromptTemplate type
Make the template on prompt a union type that outputs a list of messages if the prompt version is a "chat" type

### PR DESCRIPTION
**Pull request by anticorrelator**
resolves #5771

### CODE REVIEW

File: src/phoenix/server/api/types/PromptVersion.py [hunk 1]
```diff
@@ -29,7 +31,7 @@ class PromptVersion(Node):
     description: str
     template_type: PromptTemplateType
     template_format: PromptTemplateFormat
-    template: JSON
+    template: PromptTemplateVersion
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersion.py:34]: `template_version`?

File: app/schema.graphql [hunk 2]
```diff
@@ -1086,6 +1086,11 @@ type JSONInvocationParameter implements InvocationParameterBase {
   defaultValue: JSON
 }
 
+type JSONPromptMessageGQL {
```
Comment (axiomofjoy) [inline app/schema.graphql:1089]: Can we remove `GQL` from the type names in the GraphQL schema?

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@strawberry.enum
+class PromptMessageRole(str, Enum):
+    USER = "user"
+    SYSTEM = "system"
+    AI = "ai"  # E.g. the assistant. Normalize to AI for consistency.
+
+
+class PromptStringTemplate(BaseModel):
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersionTemplate.py:20]: I see the appeal of keeping the Pydantic models next to their corresponding GraphQL types, but will these models also be needed for the REST API?

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+class TextPromptMessage(BaseModel):
+    role: PromptMessageRole
+    content: str
+
+    def to_gql(self):
+        return TextPromptMessageGQL.from_model(self)
+
+
+class JSONPromptMessage(BaseModel):
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersionTemplate.py:32]: Do we need both to and from methods?

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    role: PromptMessageRole
+    content: JSONSerializable
+
+    def to_gql(self):
+        return JSONPromptMessageGQL.from_model(self)
+
+
+class PromptMessagesTemplateV1(BaseModel):
+    _version: str = "messages-v1"
+    template: list[Union[TextPromptMessage, JSONPromptMessage]]
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersionTemplate.py:42]: `message_templates` or `messages`?

Comment (axiomofjoy) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: `ChatTemplateV1`?

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 4 above)
Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:32]: maybe eventually but we're feeling it out

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 5 above)
Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: I think it's probably a good idea for the different template types to conform to the same interface for now

File: app/schema.graphql (hunk 2 above)
Comment (mikeldking) [reply app/schema.graphql:1089]: Agreed

File: app/schema.graphql [hunk 6]
```diff
@@ -1086,6 +1086,11 @@ type JSONInvocationParameter implements InvocationParameterBase {
...
+  role: PromptMessageRole!
```
Comment (mikeldking) [inline app/schema.graphql:1090]: I think this might be simpler to be string?

File: app/schema.graphql [hunk 7]
```diff
@@ -1395,6 +1400,17 @@ type PromptEdge {
...
 
+enum PromptMessageRole {
+  USER
+  SYSTEM
+  AI
+}
+
+type PromptMessagesTemplateV1GQL {
+  version: String!
+  template: [TextPromptMessageGQLJSONPromptMessageGQL!]!
```
Comment (mikeldking) [inline app/schema.graphql:1411]: I'm not following why this is a list?

File: app/schema.graphql [hunk 8]
```diff
@@ -1414,14 +1434,16 @@ enum PromptTemplateType {
...
+
 type PromptVersion implements Node {
   """The Globally Unique ID of this object"""
   id: GlobalID!
   user: String
   description: String!
   templateType: PromptTemplateType!
   templateFormat: PromptTemplateFormat!
-  template: JSON!
+  template: PromptTemplateVersion!
```
Comment (mikeldking) [inline app/schema.graphql:1446]: The union type should be here

File: src/phoenix/server/api/types/PromptVersion.py (hunk 1 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersion.py:34]: This is the wrong type

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 3 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:20]: Yeah they will - let's de-couple

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 5 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: This should just be a union no?

Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: ah right

Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: wait no, I think it's list[Union[...]] since there can be multiple messages

Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: they key should be `messages` that's what's confusing here

Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: ah I see, I changed it to `template` so all of our template wrappers have the same interface which I'm kind of partial to (the wrapper for the plain string template just has a `template` key)

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 3 above)
Comment (anticorrelator) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:20]: @axiomofjoy I will say decoupling these into different files might lead to some weird circular import stuff

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 5 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:42]: It shouldn't leak into the graphQL type

File: src/phoenix/server/api/types/PromptVersionTemplate.py (hunk 3 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/PromptVersionTemplate.py:20]: You can reduce the circular imports by just having the conversion code be functional in its own file and not be attached to classes

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# Part of the Phoenix PromptHub feature set
+
+from typing import Union
+
+import strawberry
+from strawberry.scalars import JSON
+
+from phoenix.server.api.helpers.prompthub.models import (
```
Comment (mikeldking) [inline src/phoenix/server/api/types/PromptVersionTemplate.py:8]:
```suggestion
from phoenix.server.api.helpers.prompts.models import (
```

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+
+@strawberry.type
+class PromptStringTemplate:
+    template: str
+
+
+PromptTemplate = strawberry.union(
+    "PromptTemplateVersion", (PromptStringTemplate, PromptChatTemplateV1)
+)
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersionTemplate.py]: Maybe just name `PromptTemplateVersion` to match the underlying GraphQL type.

File: src/phoenix/server/api/types/PromptVersionTemplate.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+@strawberry.type
+class TextPromptMessage:
+    role: PromptMessageRole
+    content: str
+
+
+@strawberry.type
+class JSONPromptMessage:
+    role: PromptMessageRole
+    content: JSON
```
Comment (axiomofjoy) [inline src/phoenix/server/api/types/PromptVersionTemplate.py:22]: Is this for tool call messages, images, etc.?

File: app/schema.graphql [hunk 12]
```diff
@@ -1783,6 +1805,13 @@ type TextChunk implements ChatCompletionSubscriptionPayload {
   content: String!
 }
 
+type TextPromptMessage {
+  role: PromptMessageRole!
+  content: String!
+}
+
+union TextPromptMessageJSONPromptMessage = TextPromptMessage | JSONPromptMessage
```
Comment (axiomofjoy) [inline app/schema.graphql:1813]: Can we name this type as `PromptMessage` or `PromptMessageTemplate`?

**Review by axiomofjoy — APPROVED** [review]

File: app/schema.graphql [hunk 13]
```diff
@@ -1377,6 +1382,11 @@ type Prompt implements Node {
   promptVersions(first: Int = 50, last: Int, after: String, before: String): PromptVersionConnection!
 }
 
+type PromptChatTemplateV1 {
```
Comment (mikeldking) [inline app/schema.graphql:1385]:
```suggestion
type PromptChatTemplate {
```

File: app/schema.graphql [hunk 14]
```diff
@@ -1377,6 +1382,11 @@ type Prompt implements Node {
...
+  version: String!
```
Comment (mikeldking) [inline app/schema.graphql:1386]:
```suggestion
  __version: String!
```

File: app/schema.graphql [hunk 15]
```diff
@@ -1395,6 +1405,12 @@ type PromptEdge {
   node: Prompt!
 }
 
+enum PromptMessageRole {
+  USER
+  SYSTEM
```
Comment (mikeldking) [inline app/schema.graphql:1410]: add a description that this includes the openAI developer role

File: app/schema.graphql [hunk 16]
```diff
@@ -1395,6 +1405,12 @@ type PromptEdge {
   node: Prompt!
 }
 
+enum PromptMessageRole {
+  USER
```
Comment (mikeldking) [inline app/schema.graphql:1409]: Missing tool

**Pull request merged**